crates/ruff_python_semantic/src/module/resolver.rs

```diff
@@ -0,0 +1,933 @@
+use std::ops::Deref;
+use std::sync::Arc;
+
+use ruff_db::file_system::{FileSystem, FileSystemPath, FileSystemPathBuf};
+use ruff_db::vfs::{file_system_path_to_file, vfs_path_to_file, VfsFile, VfsPath};
+
+use crate::module::resolver::internal::ModuleResolverSearchPaths;
+use crate::module::{
+    Module, ModuleInner, ModuleKind, ModuleName, ModuleSearchPath, ModuleSearchPathKind,
+};
+use crate::Db;
+
+const TYPESHED_STDLIB_DIRECTORY: &str = "stdlib";
+
+/// Configures the module search paths for the module resolver.
+///
+/// Must be called before calling any other module resolution functions.
+pub fn set_module_resolution_settings(db: &mut dyn Db, config: ModuleResolutionSettings) {
+    // There's no concurrency issue here because we hold a `&mut dyn Db` reference. No other
+    // thread can mutate the `Db` while we're in this call, so using `try_get` to test if
+    // the settings have already been set is safe.
+    if let Some(existing) = ModuleResolverSearchPaths::try_get(db) {
+        existing
+            .set_search_paths(db)
+            .to(config.into_ordered_search_paths());
+    } else {
+        ModuleResolverSearchPaths::new(db, config.into_ordered_search_paths());
+    }
+}
+
+/// Resolves a module name to a module.
+#[tracing::instrument(level = "debug", skip(db))]
+pub fn resolve_module(db: &dyn Db, module_name: ModuleName) -> Option<Module> {
+    let interned_name = internal::ModuleNameIngredient::new(db, module_name);
+
+    resolve_module_query(db, interned_name)
+}
+
+/// Salsa query that resolves an interned [`ModuleNameIngredient`] to a module.
+///
+/// This query should not be called directly. Instead, use [`resolve_module`]. It only exists
+/// because Salsa requires the module name to be an ingredient.
+#[salsa::tracked]
+pub(crate) fn resolve_module_query(
+    db: &dyn Db,
+    module_name: internal::ModuleNameIngredient,
+) -> Option<Module> {
+    let name = module_name.name(db);
+
+    let (search_path, module_file, kind) = resolve_name(db, name)?;
+
+    let module = Module {
+        inner: Arc::new(ModuleInner {
+            name: name.clone(),
+            kind,
+            search_path,
+            file: module_file,
+        }),
+    };
+
+    Some(module)
+}
+
+/// Resolves the module for the given path.
+///
+/// Returns `None` if the path is not a module locatable via `sys.path`.
+#[tracing::instrument(level = "debug", skip(db))]
+pub fn path_to_module(db: &dyn Db, path: &VfsPath) -> Option<Module> {
+    // It's a bit surprising to have `path_to_module` that resolves the file for `path` when the first thing that `file_to_module` does
+    // is retrieving the path from `file`. The reason for this "odd" looking API is that
+    // `file_to_module` is cached using Salsa and:
+    // * Salsa only allows ingredients as query arguments
+    // * `path` isn't a salsa ingredient, but `file` is
```

Comment:
I find this comment slightly confusing because you say that `path_to_module` resolves to a file, but the type signature says that it returns `Option<Module>`

Reply:
What I mean by `resolves` is that it performs a vfs lookup to get the file for the given path. It's not about the return type. I can try to clarify this.